README.md
# ADS

ADS is an HTTP(S) server which calculate and stores statistics data. 

https://github.com/sawrus/ads

# Features

 * Flexible way for setting server variables
 * Apply Restful way for handling the input GET requests
 * OOB support to work in SSL mode
 * Ability to Extend and change statistics data via env variables (inflight changes)
 * Ability to upload files with any format (power on cache mode for html files)
 * If cached html file was changed ie size was changed then this file will be put in NoSQL cache# Quick Start


 The typical 'Hello World" example code is:

```erlang

% application start callback
application:start(ads).

```
 
# Installation notes

  You should make the next steps:

 * Need to install openssl (>= 1.0.1) [OpenSSL](http://www.openssl.org/source/openssl-1.0.1c.tar.gz)
 * Need to install erlang (>=R14B01)
 * Need to install redis 
 * Download the last stable ADS release [ADS](https://github.com/downloads/sawrus/ads/ads_v0.1.tar.gz)
 * Unzip archive: tar -xzf ads_vMj.Mi.tar.gz
 * Run it: As daemon via (./bin/ads start) or As console via (/bin/ads console)

# Module Exports

The complete list of module exports can be found [here](https://github.com/ostinelli/misultin/tree/master/EXPORTS.md).

# Dependencies

 You will need:

 * [Erlang](http://www.erlang.org/download.html) >= R14B01
 * [Rebar](https://github.com/basho/rebar) to compile
 * [Redis](http://redis.io/download) to compile
 * [Eredis](https://github.com/wooga/eredis.git)
 * [Misultin](http://github.com/ostinelli/misultin.git)



CHANGELOG.md
# CHANGELOG

### 0.1:
 * initial release.



apps/ads/src/ads_test.erl
%%
%% ADS - Main module for testing functionality
%%
-module(ads_test).
-vsn("0.1").

%% Eunit macros
-ifdef(TEST).
-include_lib("eunit/include/eunit.hrl").
-endif.

%% Application test callbacks

-ifdef(TEST).

%% Data module tests
data_conn_test() ->
    Res = eredis:start_link(),
    ?assertMatch({ok, _}, Res),
    {ok, Conn} = Res,
    Conn.

data_redis_test() ->
    Conn = data_conn_test(),
    ?assertMatch({ok, _}, eredis:q(Conn, ["DEL", foo])),
    ?assertEqual({ok, undefined}, eredis:q(Conn, ["GET", foo])),
    ?assertEqual({ok, <<"OK">>}, eredis:q(Conn, ["SET", foo, bar])),
    ?assertEqual({ok, <<"bar">>}, eredis:q(Conn, ["GET", foo])),
    ?assertEqual({ok, <<"1">>}, eredis:q(Conn, ["DEL", foo])).

data_putget_test() ->
    Conn = data_conn_test(),
    Key = ads_util:genkey([{"A", "A"}, {"B", "B"}]),
    ?assertEqual("A:B", Key),
    ?assertEqual({ok, undefined}, ads_data:get(Key, Conn)),
    ads_data:put(Key, Key, Conn),
    {ok, Value} = ads_data:get(Key, Conn),
    ?assertEqual(<<"A:B">>, Value),
    ?assertMatch({ok, _}, eredis:q(Conn, ["DEL", Key])),
    ?assertMatch({ok, _}, eredis:q(Conn, ["DEL", "A:B:C"])),
    ?assertEqual({ok, undefined}, ads_data:get("A:B:C", Conn)),
    ads_data:put("A:B:C", "A:B:C", Conn),
    ?assertEqual({ok, <<"A:B:C">>}, ads_data:get("A:B:C", Conn)).

data_getstat_test() ->
    Conn = ads_data:open(),
    ?assertMatch({ok, _}, eredis:q(Conn, ["DEL", "K"])),
    ads_data:set_stat(0, "K", Conn),
    ?assertEqual([0, 0, 0], ads_data:get_stat("K", Conn)),
    ads_data:set_stat(1, "K", Conn),
    ?assertEqual([1, 0, 0], ads_data:get_stat("K", Conn)),
    ads_data:set_stat(2, "K", Conn),
    ?assertEqual([1, 1, 0], ads_data:get_stat("K", Conn)),
    ads_data:set_stat(3, "K", Conn),
    ?assertEqual([1, 1, 1], ads_data:get_stat("K", Conn)).

%%
%% Util module tests
%%

util_genkey_test() ->
    ?assertEqual("A", ads_util:genkey([{"A", "A"}])),
    ?assertEqual("A:B", ads_util:genkey([{"A", "A"}, {"B", "B"}])).

util_validate_test() ->
    ?assertEqual(true, ads_util:validate([{"A", "B"}], ["A"])),
    ?assertEqual(false, ads_util:validate([{"A", "B"}], ["B"])),
    ?assertEqual(true, ads_util:validate([{"A", "B"},{"A2", "B2"},{"A3","B3"}], ["A","A2","A3"])),
    ?assertEqual(false, ads_util:validate([{"A", "B"},{"A2", "B2"},{"A3","B3"}], ["A","A2","B3"])).

-endif.



apps/ads/src/ads_data.erl
%%
%% ADS - Manipulate data on NOSQL level:
%% Interface for:
%%  * Getting/Put data by KEY
%%  * Getting/Set input statistics
-module(ads_data).
-vsn("0.1").

%% API
% NoSQL callbacks
-export([get/2, put/3, open/0]).
% Statistics callbacks
-export([get_stat/2, set_stat/3]).

% includes
-include("../include/ads.hrl").

%% Application callbacks

%%
%% NoSQL functions
%%
open() ->
    Response = eredis:start_link(),
    {ok, Conn} = Response,
    Conn.

get(Key, Conn) ->
    Response = eredis:q(Conn, ["GET", Key]),
    Response.

put(Key, Value, Conn) ->
    {ok, <<"OK">>} = eredis:q(Conn, ["SET", Key, Value]).

%%
%% Stat functions
%%
inc_stat(Stat, IncNumber) -> inc_stat(Stat, IncNumber, 1).
inc_stat(Stat, IncNumber, IncValue) ->
    lists:sublist(Stat, IncNumber - 1) ++
    [lists:nth(IncNumber, Stat) + IncValue] ++
    lists:nthtail(IncNumber, Stat).


init_stat() ->
    {ok, StatUrls} = application:get_env(stat_urls),
    lists:duplicate(length(StatUrls), ?STAT_NIL).
    
get_stat(Key, Conn) ->
    {ok, Value} = get(Key, Conn),
    if
        undefined == Value ->
            Stat = init_stat();
        true ->
            Stat = binary_to_list(Value)
    end,
    Stat.

set_stat(StatNumber, Key, Conn) ->
    {ok, Stat} = get(Key, Conn),
    if
        undefined == Stat ->
            put(Key, init_stat(), Conn);
        true ->
            NewStat = inc_stat(binary_to_list(Stat), StatNumber),
            put(Key, NewStat, Conn)
    end.


apps/ads/src/ads_http.erl
%%
%% ADS - Module with call the next point:
%%  * Starting Misultin HTTP server with env parameters
%%  * Handling GET requests in loop
%%  * Stop functionality
%%
%% Usage:
%%  * ads_http:start();
%%  * ads_http:stop();
-module(ads_http).
-vsn("0.1").

% includes
-include("../include/ads.hrl").

%%% API
-export([start/0, stop/0]).

%%% HTTP Misultin callbacks

% Start <M> HTTP server
start() ->
    {ok, Port} = application:get_env(http_port),
    {ok, Folder} = application:get_env(http_folder),
    {ok, Compress} = application:get_env(http_compress),
    {ok, Ssl} = application:get_env(https_enable),
    Conn = ads_data:open(),
    ServerOptions = [
        {port, Port},
        {access_log, fun(AccessInfo) -> access_log(AccessInfo) end},
        {compress, Compress},
        {static, Folder},
        {loop, fun(Req) -> 
            try
                ads_req:handle(Req, Conn) 
            catch        
                Exception : Reason -> 
                    ETempl = "Exception: ~p~nReason: ~p~nStacktrace: ~p",
                    EList  = [Exception, Reason, erlang:get_stacktrace()],
                    ?LOG_DEBUG(ETempl, EList),
                    Message = io_lib:format(ETempl, EList),
                    Req:respond(500, [{"Content-Type", "text/plain"}], Message)
            end
        end
        }
    ],
    ?LOG_DEBUG("Start with ServerOptions: ~p", [ServerOptions]),
    if 
        false == Ssl ->
            misultin:start_link(ServerOptions);
        true ->
            {ok, SslOptions} = application:get_env(https_options),
            misultin:start_link(lists:append(ServerOptions, [{ssl, SslOptions}]))
    end.

% callback on access log
access_log({PeerAddr, DateTime, RequestLine, HttpCode, ContentLength}) ->
    io:format("~s - - [~s] \"~s\" ~p ~p~n", [PeerAddr, DateTime, RequestLine, HttpCode, ContentLength]).

% Stop <M> HTTP server
stop() ->
    misultin:stop().


apps/ads/src/ads_util.erl
%%
%% ADS - Ad server utility
%%
-module(ads_util).
-vsn("0.1").

%% includes
-include("../include/ads.hrl").
-include_lib("kernel/include/file.hrl").

%% API
-export([genkey/1, genkey/2]).
-export([validate/2]).
-export([get_mtime/1]).
-export([index_of/2]).
-export([sub/3, gsub/3]).
-export([build_page/2, get_page/2]).

%% Application callbacks

%% @doc Generate Key based on Args proplist
%% Usage:
%% ads_util:genkey([{"A", "B"}, {"C", "D"}]) returns string "B:D"
-spec genkey(TupleList::[tuple()]) -> string().
genkey(TupleList) -> genkey(TupleList, ?STAT_SEPARATOR).

-spec genkey(TupleList::[tuple()], Sep::string()) -> string().
genkey([{_, H} | T], Sep) ->
    H ++ lists:append([Sep ++ Key || {_, Key} <- T]).

%% @doc Validate that keys of proplist equals the second argument of this function.
%% Usage:
%% ads_util:validate([{A, B}, {C, D}], [A, C]) returns true
%% ads_util:validate([{A, B}, {C, D}], [B, D]) returns false
-spec validate(PropList, ValList) -> ValResult when
    PropList::[tuple()],
    ValList::[string()],
    ValResult::boolean.
validate([], []) -> true;
validate(L1, L2) when length(L1) =/= length(L2) -> false;
validate([{A, _} | T1], [E | T2]) -> 
    R = string:equal(A, E),
    if 
        false == R -> false;
        true -> validate(T1,T2)    
    end.

%% @doc Getting modification time in seconds of defined File
%% Usage:
%% ads_util:get_mtime(File)
-spec get_mtime(FilePath::string()) -> integer().
get_mtime(FilePath) ->
    {ok, FileInfo} = file:read_file_info(FilePath),
    {Date, Time} = FileInfo#file_info.mtime,
    {YYYY,MM,DD} = Date, {HH,Mi,SS} = Time,
    MTime = YYYY + MM + DD + HH + Mi + SS,
    MTime.

index_of(Item, List) -> index_of(Item, List, 1).
index_of(_, [], _)  -> -1;
index_of(Item, [Item|_], Index) -> Index;
index_of(Item, [_|Tl], Index) -> index_of(Item, Tl, Index+1).

-spec sub(Str::string, Old::string(), New::string()) -> string().
sub(Str, Old, New) ->
    RegExp = "\\Q"++Old++"\\E",
    re:replace(Str,RegExp,New,[multiline, {return, list}]).

-spec gsub(Str::string, Old::string(), New::string()) -> string().
gsub(Str, Old, New) ->
    RegExp = "\\Q"++Old++"\\E",
    re:replace(Str,RegExp,New,[global, multiline, {return, list}]).

save_file(FilePath, Url, Conn) ->
     {ok, BinaryFileContent} = file:read_file(FilePath),
     FileModifiedTime = io_lib:format("~p", [ads_util:get_mtime(FilePath)]),
     FileContent = binary_to_list(BinaryFileContent),
     ads_data:put(Url, string:join([FileModifiedTime, FileContent], ?HTML_SEPARATOR), Conn),
     get_page(Url, Conn).
     
get_page("", _) -> "";
get_page(PageName, Conn) ->
    {ok, Folder} = application:get_env(http_folder),
    FilePath = filename:join([Folder, PageName]),
    FileExist = filelib:is_file(FilePath),
    Result = ads_data:get(PageName, Conn),
    if 
        false == FileExist -> PageName;
        true ->
        case Result of 
            {ok, undefined} -> 
                save_file(FilePath, PageName, Conn);
            {ok, BinaryFileContent} -> 
                ActualModifiedTime = integer_to_list(ads_util:get_mtime(FilePath)),
                [CachedModifiedTime | FileContent] = string:tokens(binary_to_list(BinaryFileContent), ?HTML_SEPARATOR),
                if 
                    ActualModifiedTime =/= CachedModifiedTime ->
                        save_file(FilePath, PageName, Conn);
                    true ->
                        build_page(FileContent, Conn)
                end;
            _ -> PageName
        end
    end.

build_page([], _) -> "";
build_page([PagePart | T], Conn) ->
    IsHtmlPart = lists:suffix(".html", PagePart),
    if 
        false == IsHtmlPart ->
            Result = PagePart;
        true ->
            Result = get_page(PagePart, Conn)
    end,
    Result ++ build_page(T, Conn).


apps/ads/src/ads_req.erl
%%
%% ADS - Handle request module
%%
-module(ads_req).
-vsn("0.1").

% includes
-include("../include/ads.hrl").
-include_lib("kernel/include/file.hrl").

% API
-export([handle/2]).

%% @doc Handle HTTP request callbacks
handle(Req, Conn) ->
    % get params depending on method
    Method = Req:get(method),
    case Method of
        'GET' ->
            Args = Req:parse_qs();
        'POST' ->
            Args = Req:parse_post()
    end,
    Uri = Req:resource([lowercase, urldecode]),
    % Handle request by parameters
    {ok, StatUrls} = application:get_env(stat_urls),
    {ok, ReportUrls} = application:get_env(report_urls),
    {ok, ConfigUrl} = application:get_env(config_url),
    {ok, HomeUrl} = application:get_env(home_url),
    {ok, UploadUrl} = application:get_env(upload_url),
    StatIndex = ads_util:index_of(Uri, StatUrls),
    ReportIndex = ads_util:index_of(Uri, ReportUrls),
    if 
        StatIndex > 0 ->
            calc_stat(StatIndex, Args, Req, Conn);
        ReportIndex > 0 ->
            prepare_report(Args, Req, Conn);
        ConfigUrl == Uri ->
            prepare_config(Args, Req, Conn);
        HomeUrl == Uri ->
            handle_home(Req);
        [] == Uri ->
            handle_home(Req);
        UploadUrl == Uri ->
            handle_upload(Method, Args, Req);
        true ->
            handle(Uri, Req, Conn)        
    end.

calc_stat(IncNumber, Args, Req, Conn) ->
    {ok, StatKeys} = application:get_env(stat_keys),
    ValidateResult = ads_util:validate(Args, StatKeys), 
    if 
        false == ValidateResult -> Req:respond(400); 
        true ->
            ads_data:set_stat(IncNumber, ads_util:genkey(Args), Conn),
            Req:respond(200)
    end.

prepare_report(Args, Req, Conn) ->
    {ok, StatKeys} = application:get_env(stat_keys),
    ValidateResult = ads_util:validate(Args, StatKeys), 
    if 
        false == ValidateResult -> Req:respond(400); 
        true ->
            Key = ads_util:genkey(Args),
            Value = ads_data:get_stat(Key, Conn),
            JSONTemplate = "{\"~s\":\"~p\"}",
            JSON = io_lib:format(JSONTemplate, [Key, Value]),
            Req:ok([{"Content-Type", "application/json"}], JSON)
    end.

prepare_config(Key) ->
    {H, M, S} = time(),
    Time  = io_lib:format('~2..0b:~2..0b:~2..0b', [H, M, S]),
    Extra = [{"Key", Key}, {"Time", Time}],
    BuildJSON = fun({Param, Value}, Acc) ->
        [lists:flatten(
            io_lib:format("{\"~s\":\"~s\"},\n", [Param, Value])) | Acc
        ]
    end,
    ExtraJson = lists:flatten(lists:reverse(lists:foldl(BuildJSON, [], Extra))),
    ExtraJson.

save_file(FilePath, Url, Req, Conn) ->
    {ok, BinaryFileContent} = file:read_file(FilePath),
    FileModifiedTime = io_lib:format("~p", [ads_util:get_mtime(FilePath)]),
    FileContent = binary_to_list(BinaryFileContent),
    ads_data:put(Url, string:join([FileModifiedTime, FileContent], ?HTML_SEPARATOR), Conn),
    Req:ok([{"Content-Type", "text/html"}], ads_util:build_page(string:tokens(FileContent, ?HTML_SEPARATOR), Conn)).

prepare_config(Args, Req, Conn) ->
    {ok, ConfigKeys} = application:get_env(config_keys),
    ValidateResult = ads_util:validate(Args, ConfigKeys), 
    Key = ads_util:genkey(Args),
    {ok, Value} = ads_data:get(Key, Conn),
    if
        false == ValidateResult ->
            Req:respond(400);
        undefined == Value ->
            %% ! test mode !
            NewValue = prepare_config(Key),
            ads_data:put(Key, NewValue, Conn),
            Req:ok([{"Content-Type", "application/json"}], NewValue);
        true ->
            Req:ok([{"Content-Type", "application/json"}], Value)
    end.

build_hrefs([], _, _) -> "";
build_hrefs([H | T], Keys, Def) ->
    build_href(H, Keys, Def) ++ build_hrefs(T, Keys, Def).
    
build_href(Url, Keys, Def) ->
    Href = "/" ++ 
        string:join(Url, "/") ++ "?" ++ 
        string:join(Keys, "=" ++ Def ++ "&") ++ "=" ++ Def,
    "<li><a href=\""++ Href ++ "\">" ++ Href ++ "</a></li>\n".
     
handle_home(Req) ->
    {ok, StatUrls} = application:get_env(stat_urls),
    {ok, StatKeys} = application:get_env(stat_keys),
    {ok, ReportUrls} = application:get_env(report_urls),
    {ok, ConfigUrl} = application:get_env(config_url),
    {ok, ConfigKeys} = application:get_env(config_keys),
    {ok, UploadUrl} = application:get_env(upload_url),
    StatDef = io_lib:format("~p", [random:uniform(51) + 24]),
    ConfigDef = io_lib:format("~p", [random:uniform(75) + 51]),
    ReportHref = build_hrefs(ReportUrls, StatKeys, StatDef),
    StatHref = build_hrefs(StatUrls, StatKeys, StatDef),
    ConfigHref = build_href(ConfigUrl, ConfigKeys, ConfigDef),
    UploadHref = "/" ++ string:join(UploadUrl, "/"),
    Req:ok([{"Content-Type", "text/html"}], 
["<ul>\n", ReportHref, "\n", StatHref, "\n", ConfigHref,"\n
  <li><a href=\"", UploadHref, "\">", UploadHref, "</a</li></ul>"]).

handle_upload(Method, Args, Req) ->
    if 
        'GET' == Method ->
            Req:ok([{"Content-Type", "text/html"}], 
                ["<html><head><title>File Upload</title></head><body>
                  <form action=\"/upload\" method=\"POST\" enctype=\"multipart/form-data\">
                  <input type=\"file\" name=\"file\">
                  <input type=\"submit\">
                  </form></body></html>"]);
        true ->
        case Args of
            [{_Tag, Attributes, FileData}] ->
                % build destination file path
                {ok, DestPath} = application:get_env(http_folder),
                FileName = misultin_utility:get_key_value("filename", Attributes),
                DestFile = DestPath ++ "/" ++ FileName,
                % save file
                case file:write_file(DestFile, FileData) of
                    ok ->
                        Req:file(DestFile);
                    {error, _Reason} ->
                        Req:respond(500)
                end;
            _ ->
                Req:respond(500)
        end
    end.

handle(Url, Req, Conn) ->
    {ok, Folder} = application:get_env(http_folder),
    FullUrl = filename:join(Url),
    FilePath = filename:join([Folder, FullUrl]),
    {ok, BinaryFileContent} = ads_data:get(FullUrl, Conn),
    FileExist = filelib:is_file(FilePath), 
    HtmlFile = lists:suffix(".html", FilePath),
    if
        false == FileExist ->
            Req:respond(404);
        false == HtmlFile ->
            Req:file(FilePath);
        undefined == BinaryFileContent->
            save_file(FilePath, FullUrl, Req, Conn);
        true ->                            
            ActualModifiedTime = integer_to_list(ads_util:get_mtime(FilePath)),
            [CachedModifiedTime | FileContent] = string:tokens(binary_to_list(BinaryFileContent), ?HTML_SEPARATOR),
            if 
                ActualModifiedTime =/= CachedModifiedTime  ->
                    save_file(FilePath, FullUrl, Req, Conn);
                true ->
                    Req:ok([{"Content-Type", "text/html"}], ads_util:build_page(FileContent, Conn))
            end
    end.

